New lambda provider v2 (#532)
localstack/docs: content/en/user-guide/tools/lambda-tools/hot-reloading/index.md

Please change your deployment configuration accordingly because the old value is an invalid bucket name.
The configuration `BUCKET_MARKER_LOCAL` is still supported.

More information about the new lambda provider is available under [Lambda Provider Behavioral Changes]({{< ref "lambda-provider-v2" >}}).
{{< /alert >}}

## Covered Topics

* [Hot Reloading Behavior](#hot-reloading-behavior)
* [Application Configuration Examples](#application-configuration-examples):
  * [Hot reloading for JVM Lambdas](#hot-reloading-for-jvm-lambdas)
  * [Hot reloading for Python Lambdas](#hot-reloading-for-python-lambdas)
  * Debugging Nodejs lambdas (under development)
* [Deployment Configuration Examples](#deployment-configuration-examples):
  * [Serverless Framework Configuration](#serverless-framework-configuration)
  * [AWS Cloud Development Kit (CDK) Configuration](#aws-cloud-development-kit-cdk-configuration)
  * [Terraform Configuration](#terraform-configuration)
* [Useful Links](#useful-links)

## Hot Reloading Behavior

**Delay in code change detection:**
It can take up to 700ms to detect code changes.
In the meantime, invocations still execute the former code.

> Hot reloading triggers after 500ms without changes, can take up to 200ms until the change is notified.

Reviewer: ```suggestion
Hot reloading triggers after 500ms without changes, and it can take up to an additional 200ms until the reloaded code is in effect.
```

**Runtime restart after each code change:**
The runtime inside the container is restarted after every code change.

> Therefore, the handler function re-executes any initialization code *outside* after every code change.

Reviewer: ```suggestion
During the runtime restart, the handler function re-executes any initialization code *outside* your handler function.
```

Author: updated and synced with lambda v2 page

> However, the container itself does not restart.
> Therefore, filesystem changes persist between code changes for invocations dispatched to the same container.

Reviewer: ```suggestion
The container itself is not restarted.
Therefore, filesystem changes persist between code changes for invocations dispatched to the same container.
```

$ git clone git@github.com:awsdocs/aws-doc-sdk-examples.git
{{< / command >}}

#### Creating the Lambda Function

To create the Lambda function, you just need to take care of two things:
1. Deploy via an S3 Bucket. You need to use the magic variable `hot-reload` as the bucket.
2. Set the S3 key to the path of the directory your lambda function resides in.
   The handler is then referenced by the filename of your lambda code and the function in that code that needs to be invoked.

So, using the AWS example, this would be:

{{< command >}}
$ awslocal lambda create-function --function-name my-cool-local-function \
    --code S3Bucket="hot-reload",S3Key="/tmp/aws-doc-sdk-examples/python/example_code/lambda/boto_client_examples" \
    --handler lambda_handler_basic.lambda_handler \
    --runtime python3.8 \

>     --role arn:aws:iam::000000000000:role/lambda-role

Author: Should we keep the sample ARN region-agnostic or add `us-east-1`?

(I generally prefer `123456789012` to clarify the number of digits but 12x `0` seems more consistent with LocalStack returning it and assuming it as role)

Author: IAM is account-global without region. => Correct as now
